enterprise/server/cmd/ci_runner/main.go

```diff
@@ -0,0 +1,643 @@
+
+	// NOTE: In this func we return immediately with an error of nil if event publishing fails,
+	// because that error is instead surfaced in the caller func when calling
+	// `buildEventPublisher.Wait()`
+
+	bep := ar.bep
+
+	startedEvent := &bespb.BuildEvent{
+		Id: &bespb.BuildEventId{Id: &bespb.BuildEventId_Started{Started: &bespb.BuildEventId_BuildStartedId{}}},
+		Children: []*bespb.BuildEventId{
+			{Id: &bespb.BuildEventId_Progress{Progress: &bespb.BuildEventId_ProgressId{OpaqueCount: 0}}},
+			{Id: &bespb.BuildEventId_WorkspaceStatus{WorkspaceStatus: &bespb.BuildEventId_WorkspaceStatusId{}}},
+			{Id: &bespb.BuildEventId_BuildFinished{BuildFinished: &bespb.BuildEventId_BuildFinishedId{}}},
+		},
+		Payload: &bespb.BuildEvent_Started{Started: &bespb.BuildStarted{
+			Uuid:            ar.bep.streamID.InvocationId,
+			StartTimeMillis: time.Now().UnixNano() / int64(time.Millisecond),
+		}},
+	}
+	if err := bep.Publish(startedEvent); err != nil {
+		return nil
+	}
+	if err := ar.flushProgress(); err != nil {
+		return nil
+	}
+	workspaceStatusEvent := &bespb.BuildEvent{
+		Id: &bespb.BuildEventId{Id: &bespb.BuildEventId_WorkspaceStatus{WorkspaceStatus: &bespb.BuildEventId_WorkspaceStatusId{}}},
+		Payload: &bespb.BuildEvent_WorkspaceStatus{WorkspaceStatus: &bespb.WorkspaceStatus{
+			Item: []*bespb.WorkspaceStatus_Item{
+				{Key: "BUILD_USER", Value: ar.username},
+				{Key: "BUILD_HOST", Value: ar.hostname},
+				{Key: "REPO_URL", Value: *repoURL},
+				{Key: "COMMIT_SHA", Value: *commitSHA},
+				{Key: "GIT_TREE_STATUS", Value: "Clean"},
+				// TODO: Populate GIT_BRANCH. Can't source this from the `trigger_branch` flag
+				// in the PR case, because that refers to the branch into which the PR would be
+				// merged, which doesn't reflect the currently checked out branch.
+			},
+		}},
+	}
+	if err := bep.Publish(workspaceStatusEvent); err != nil {
+		return nil
+	}
+
+	ar.log.writeListener = func() {
+		// No need to lock the buffer here (read-only operation)
+		if size := ar.log.Buffer.Len(); size >= progressFlushThresholdBytes {
+			ar.flushProgress() // ignore error
+		}
+	}
+	stopFlushingProgress := ar.startBackgroundProgressFlush()
+	defer stopFlushingProgress()
+
+	for _, bazelCmd := range ar.action.BazelCommands {
+		args, err := bazelArgs(bazelCmd)
+		if err != nil {
+			return fmt.Errorf("failed to parse bazel command: %s", err)
+		}
+		ar.printCommandLine(args)
+		err = runCommand(ctx, "bazelisk", args /*env=*/, nil, ar.log)
+		if exitCode := getExitCode(err); exitCode != noExitCode {
+			// not using aurora here because our client-side library doesn't
+			// support the types of escape sequences required for aurora's
+			// fancy "grayscale" colors.
+			ar.log.Printf("\033[90m(command exited with code %d)\033[0m", exitCode)
+		}
+		// If the command failed, report its progress before returning.
+		if err := ar.flushProgress(); err != nil {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
```

Comment:
i don't quite follow this bit -- seems like the first branch will override err? Also do you really want to return nil if err != nil?